statement.go

```diff
@@ -515,30 +532,88 @@ func (s *ShowTablesStatement) Execute(ctx context.Context, session *Session) (*R
 
 type ExplainStatement struct {
 	Explain string
+	IsDML   bool
 }
 
+// Execute processes `EXPLAIN` statement for queries and DMLs.
 func (s *ExplainStatement) Execute(ctx context.Context, session *Session) (*Result, error) {
-	stmt := spanner.NewStatement(s.Explain)
-	queryPlan, err := session.client.Single().AnalyzeQuery(ctx, stmt)
+	queryPlan, timestamp, _, err := runAnalyzeQuery(ctx, session, spanner.NewStatement(s.Explain), s.IsDML)
 	if err != nil {
 		return nil, err
 	}
 
+	if queryPlan == nil {
+		return nil, errors.New("EXPLAIN statement is not supported for Cloud Spanner Emulator.")
+	}
+
+	var rowCount int
+	if !s.IsDML {
+		rowCount = 1
+	}
+
 	rows, predicates, err := processPlanWithoutStats(queryPlan)
 	if err != nil {
 		return nil, err
 	}
 
 	result := &Result{
+		IsMutation:   s.IsDML,
 		ColumnNames:  explainColumnNames,
-		AffectedRows: 1,
+		AffectedRows: rowCount,
 		Rows:         rows,
+		Timestamp:    timestamp,
 		Predicates:   predicates,
 	}
 
 	return result, nil
 }
 
+type DescribeStatement struct {
+	Statement string
+	IsDML     bool
+}
+
+// Execute processes `DESCRIBE` statement for queries and DMLs.
+func (s *DescribeStatement) Execute(ctx context.Context, session *Session) (*Result, error) {
+	_, timestamp, metadata, err := runAnalyzeQuery(ctx, session, spanner.NewStatement(s.Statement), s.IsDML)
+	if err != nil {
+		return nil, err
+	}
+
+	var rowCount int
+	if !s.IsDML {
+		rowCount = 1
+	}
```

Comment:
What is the expected result from this?

I tested and it showed 0 rows for DML and 1 row for Query.

```
spanner> describe insert into Players (PlayerId, PlayerName) values (5, 'c') then return *;
+-------------+-------------+
| Column_Name | Column_Type |
+-------------+-------------+
| PlayerId    | INT64       |
| PlayerName  | STRING      |
+-------------+-------------+
Query OK, 0 rows affected (0.45 sec)

spanner> describe select * from Players;
+-------------+-------------+
| Column_Name | Column_Type |
+-------------+-------------+
| PlayerId    | INT64       |
| PlayerName  | STRING      |
+-------------+-------------+
1 rows in set (0.18 sec)
```

Both results show two rows, so isn't it show `2 rows in set` in both results?

Reply:
It is fair comment. 
It is a legacy that current code always prints `0 rows affected` in `ExplainDmlStatement` and `1 rows in set` in `ExplainStatement`.

Personally I think it is no strong reason to show `RowCount` when statements don't execute any actual query.

Reply:
Fixed https://github.com/cloudspannerecosystem/spanner-cli/pull/191/commits/4b847eb21bc1a971466f2f49d30b63dac4731b8d
```
spanner> EXPLAIN SELECT * FROM Singers;
+----+-------------------------------------------------------------------------------+
| ID | Query_Execution_Plan                                                          |
+----+-------------------------------------------------------------------------------+
|  0 | Distributed Union (distribution_table: Singers, split_ranges_aligned: false)  |
|  1 | +- Local Distributed Union                                                    |
|  2 |    +- Serialize Result                                                        |
|  3 |       +- Table Scan (Full scan: true, Table: Singers, scan_method: Automatic) |
+----+-------------------------------------------------------------------------------+
4 rows in set (0.15 sec)

spanner> EXPLAIN DELETE FROM Singers WHERE TRUE THEN RETURN WITH ACTION *;
+----+----------------------------------------------------------------------------------+
| ID | Query_Execution_Plan                                                             |
+----+----------------------------------------------------------------------------------+
|  0 | Apply Mutations (operation_type: DELETE, table: Singers)                         |
|  1 | +- Distributed Union (distribution_table: Singers, split_ranges_aligned: false)  |
|  2 |    +- Local Distributed Union                                                    |
|  3 |       +- Serialize Result                                                        |
|  4 |          +- Table Scan (Full scan: true, Table: Singers, scan_method: Automatic) |
+----+----------------------------------------------------------------------------------+
5 rows in set (0.45 sec)
timestamp:            2024-10-15T00:06:48.154289+09:00


spanner> DESCRIBE SELECT * FROM Singers;
+-------------+-------------+
| Column_Name | Column_Type |
+-------------+-------------+
| SingerId    | INT64       |
| FirstName   | STRING      |
| LastName    | STRING      |
| SingerInfo  | BYTES       |
| BirthDate   | DATE        |
+-------------+-------------+
5 rows in set (0.14 sec)

spanner> DESCRIBE DELETE FROM Singers WHERE TRUE THEN RETURN WITH ACTION *;
+-------------+-------------+
| Column_Name | Column_Type |
+-------------+-------------+
| SingerId    | INT64       |
| FirstName   | STRING      |
| LastName    | STRING      |
| SingerInfo  | BYTES       |
| BirthDate   | DATE        |
| ACTION      | STRING      |
+-------------+-------------+
6 rows in set (0.43 sec)
timestamp:            2024-10-15T00:06:00.151541+09:00

```

Reply:
Thanks!